### ISSUE TITLE
100% Discount cart rule results in Tax invoiced

### ISSUE THREAD

**Issue by OldPlanets** [magento/magento2#30853]
### Preconditions (*)

1. Magento 2.3.6
2. Magento 2.4-develop
3. Free shipping is enabled

### Steps to reproduce (*)

1. Navigate to Stores / Settings /  Configuration / Sales / Tax
2. Set Catalog Prices and Apply Discount on Prices to Including Tax 
![image](https://user-images.githubusercontent.com/51681379/98915346-76374d80-24d2-11eb-99bf-ad73121efc31.png)
3. Navigate to Sores / Tax Rules / Add New Tax Rule 
4. Create a rule for all USA with 10% tax rate and apply it 
5. Navigate to Marketing / Cart Price Rules and Add New Rule
6. Create a rule with 100% discount for all users 
![image](https://user-images.githubusercontent.com/51681379/98915438-9535df80-24d2-11eb-93fc-46b63ae0ece2.png)
7. Make an order on Storefront:
- choose Free Shipping
- apply coupon code 
8. Navigate to Sales / Orders  and open your order
9. Create an invoice for the order and open it  

### Expected result (*)

1. An invoice is created over 0,00$

### Actual result (*)

1. An invoice with the Grand Total of the tax amount is created
![image](https://user-images.githubusercontent.com/51681379/98917476-19896200-24d5-11eb-8376-f02d5fef52be.png)
![Peek 2020-11-12 10-50](https://user-images.githubusercontent.com/51681379/98917506-227a3380-24d5-11eb-9663-da9b6426a329.gif)

I'm suspect this bug was introduced in this commit (MC-35675):
https://github.com/magento/magento2/commit/2dbab41631cce46be4a27dcbf10a7b172be61f63

It doesn't allow GrandTotals to go below zero, but the Tax Totals will add the discounted Tax, expecting that it was deducted before, resulting in a Invoice which is never smaller than the tax amount (instead of being 0)

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Delta 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-39052` were created

**Issue Available**: @engcom-Delta, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by paulvandermeijs** [comment]
I noticed in `app/code/Magento/Sales/Model/Order/Invoice/Total/Tax.php` that the discount tax compensation is added to the grand total; if I'm right that is the amount that is removed from the tax by a discount. In that case shouldn't it be excluded from the grand total?

https://github.com/magento/magento2/blob/445b0f1a3d6a91b5da32d311077c2112ef0b1503/app/code/Magento/Sales/Model/Order/Invoice/Total/Tax.php#L121-L122

**Comment by ThomasNegeli** [comment]
@paulvandermeijs 
Until 2.2.11 the $invoice->getGrandTotal() was a negative value in that case.
In 2.3 something happenend so that the $invoice->getGrandTotal() returns 0 (which is logically right, but wrong when adding tax compensation).

**Comment by chequille** [comment]
Hi All,
this is not only  tax problem, this is  general problem with generating the invoice if there is discount applied. If there is no discount, the invoice is correct.
The new code from the classes in Magento_Sales/Module/Order/Invoice/Total is not working as expected.

As I think it was correct within Magento 2.3.5, I made a new module overriding the classes Tax.php, Subtotal.php, Shippment.php and Discount.php and took the code from Magento 2.3.5

The result of it is, that everything works as it should.

Therefore, there must be some bugs in the new classes. 

As this is a mandatory and needed functionality, please be so kind and fix this as fast as possible and especially provide a patch because waiting for releases might be to long.

Thanks and best regards,
Jürgen

**Comment by chequille** [comment]
Hi
additional information about what I found on this issue.
As I said in previous comment, I override all classes in Magento_Sales/Module/Order/Invoice/Total and used the code from 2.3.5
I compared the code and found, that the only real difference is in DIscount.php and this is causing the problem.

```
        $grandTotal = $invoice->getGrandTotal() - $totalDiscountAmount < 0.0001
        ? 0 : $invoice->getGrandTotal() - $totalDiscountAmount;
        $baseGrandTotal = $invoice->getBaseGrandTotal() - $baseTotalDiscountAmount < 0.0001
        ? 0 : $invoice->getBaseGrandTotal() - $baseTotalDiscountAmount;
        
        $invoice->setGrandTotal($grandTotal);
        $invoice->setBaseGrandTotal($baseGrandTotal);

```
The invoice is wrong because $invoice->getGrandTotal() contains the amount without tax, but $totalDiscountAmount is with tax.
Therefore, as long as GrandTotal is less than TotalDiscountAmount it is getting wrong. 

Using the code from 2.3.5 at that point. Invoice are correct again.

```
    	$invoice->setDiscountAmount(-$totalDiscountAmount);
    	$invoice->setBaseDiscountAmount(-$baseTotalDiscountAmount);
    	
    	$invoice->setGrandTotal($invoice->getGrandTotal() - $totalDiscountAmount);
    	$invoice->setBaseGrandTotal($invoice->getBaseGrandTotal() - $baseTotalDiscountAmount);

```

I hope this helps to fix the this bug.

Best regards,
Chequille

**Comment by engcom-Hotel** [comment]
@magento I am working on this

**Comment by angelomaragna** [comment]
I might guess that this is of big concerns when discount is greater than 50%, as, if the GrandTotal gest negative along the overall computation, a recent Magento update f**s it up.

Thanks to @chequille I've narrowed it down to this commit, which is completely wrong, and inside Magento since 2.3.6

https://github.com/magento/magento2/commit/c710c9b38e0bcfbaedc6114bc5b4b9e62df95bbf

**The grand total is supposed to be negative** as it's not concern of the discounting class to ensure it is greater than zero, as that is later used to calculate correct amounts.

Statistically speaking, we've never had problems up to 2.3.5, and now both 2.4.1 and 2.3.6 reports computational problems.

Quick solution for us, it's been patching the file directly, as the blame show no further modification was applied to it: https://github.com/magento/magento2/blame/2.4.1/app/code/Magento/Sales/Model/Order/Invoice/Total/Discount.php

Here the patch for magento/module-sales:

```patch
--- Model/Order/Invoice/Total/Discount.php
+++ Model/Order/Invoice/Total/Discount.php
@@ -71,12 +71,12 @@
         $invoice->setDiscountAmount(-$totalDiscountAmount);
         $invoice->setBaseDiscountAmount(-$baseTotalDiscountAmount);
 
-        $grandTotal = $invoice->getGrandTotal() - $totalDiscountAmount < 0.0001
-            ? 0 : $invoice->getGrandTotal() - $totalDiscountAmount;
-        $baseGrandTotal = $invoice->getBaseGrandTotal() - $baseTotalDiscountAmount < 0.0001
-            ? 0 : $invoice->getBaseGrandTotal() - $baseTotalDiscountAmount;
-        $invoice->setGrandTotal($grandTotal);
-        $invoice->setBaseGrandTotal($baseGrandTotal);
+        // Lines reverted from 2.3.5 as https://github.com/magento/magento2/commit/c710c9b38e0bcfbaedc6114bc5b4b9e62df95bbf
+        // bogs it up. Ref: https://github.com/magento/magento2/issues/30853
+        //
+        $invoice->setGrandTotal($invoice->getGrandTotal() - $totalDiscountAmount);
+        $invoice->setBaseGrandTotal($invoice->getBaseGrandTotal() - $baseTotalDiscountAmount);
+
         return $this;
     }

```


Hope this helps to better identify and put a containment to the original problem (which is how can these commits occur in the first place...)

**Comment by barryvdh** [comment]
@angelomaragna Thanks, I will try this patch. The solution in https://github.com/magento/magento2/pull/31556 does not work for us unfortunately. 
The issue is indeed also present with other values then 100%, eg we had weird results with 99% also.

**Comment by HenKun** [comment]
@engcom-Hotel Is there any ETA or progress?

**Comment by HenKun** [comment]
I can confirm the proposed change in https://github.com/magento/magento2/issues/30853#issuecomment-764582933 works for discounts of 0%, 40%, 50%, 90% and 100% on M2.4.1. While the given patch could not be applied in my case I created a new (identical) one, see https://github.com/magento/magento2/issues/32270#issuecomment-795232668

The here mentioned fix in PR #31556 does work for discounts of 100%, however, not for discounts less than 100% and larger than 50%. (Tested on M2.4.1)

**Comment by DJB95** [comment]
@engcom-Hotel is there any ETA on this issue? Looking forward to hearing from you!

**Comment by hostep** [comment]
This is probably fixed by: [MC-41438: 100% Discount cart rule results in Tax invoiced](https://github.com/magento/magento2/commit/533ec2eb46c91b79b7a99ba4a6a164e2f6e3d7f1), which I'm guessing is going to be included in Magento 2.4.3

Maybe somebody can confirm this?
